app/src/main/java/me/hufman/androidautoidrive/music/spotify/SpotifyWebApi.kt

```diff
@@ -0,0 +1,249 @@
+import me.hufman.androidautoidrive.R
+import me.hufman.androidautoidrive.music.controllers.SpotifyAppController
+import me.hufman.androidautoidrive.phoneui.SpotifyAuthorizationActivity
+import net.openid.appauth.*
+
+/**
+ * Handles the logic around the Spotify Web API. This class is a singleton to prevent issues of having
+ * multiple [SpotifyClientAPI]s.
+ */
+class SpotifyWebApi private constructor(val context: Context) {
+	companion object {
+		const val TAG = "SpotifyWebApi"
+		const val NOTIFICATION_CHANNEL_ID = "SpotifyAuthorization"
+		const val NOTIFICATION_CHANNEL_NAME = "Spotify Authorization"
+		const val NOTIFICATION_REQ_ID = 56
+
+		private var webApiInstance: SpotifyWebApi? = null
+
+		/**
+		 * Retrieves the current [SpotifyWebApi] instance creating one if it there are no instances
+		 * currently running.
+		 */
+		fun getInstance(context: Context): SpotifyWebApi {
+			if (webApiInstance == null) {
+				webApiInstance = SpotifyWebApi(context)
+			}
+			return webApiInstance as SpotifyWebApi
+		}
+	}
+
+	// will be null if the user is not authenticated
+	private var webApi: SpotifyClientAPI? = null
+
+	private val authStateManager: SpotifyAuthStateManager
+	private val appSettingsViewer: AppSettingsViewer
+	private var getLikedSongsAttempted: Boolean = false
+	private var spotifyAppControllerCaller: SpotifyAppController? = null
+	var isUsingSpotify: Boolean = false
+
+	init {
+		Log.d(TAG, "Initializing for the first time")
+		authStateManager = SpotifyAuthStateManager(context)
+		appSettingsViewer = AppSettingsViewer()
+	}
+
+	/**
+	 * Sets the Liked Songs flag to false and sets the [SpotifyAppController] caller to null.
+	 */
+	fun clearGetLikedSongsAttemptedFlag() {
+		getLikedSongsAttempted = false
+		spotifyAppControllerCaller = null
+	}
+
+	fun getLikedSongs(spotifyAppController: SpotifyAppController): List<SpotifyMusicMetadata>?  {
+		if (webApi == null) {
+			getLikedSongsAttempted = true
+			spotifyAppControllerCaller = spotifyAppController
+			return emptyList()
+		}
+		try {
+			val likedSongs = webApi?.library?.getSavedTracks(50)?.getAllItemsNotNull()?.complete()
+			return likedSongs?.map {
+				val track = it.track
+				val mediaId = track.uri.uri
+				val artists = track.artists.map { it.name }.joinToString(", ")
+				val album = track.album
+				val coverArtIndex = album.images.indexOfFirst { it.height == 300 }
+				val imageUrl = album.images[coverArtIndex].url
+				val coverArtCode = imageUrl.substring(imageUrl.lastIndexOf("/")+1)
+				val coverArtUri = "spotify:image:$coverArtCode"
+				SpotifyMusicMetadata(spotifyAppController, mediaId, mediaId.hashCode().toLong(), coverArtUri, artists, album.name, track.name)
+			}
+		} catch (e: SpotifyException.AuthenticationException) {
+			Log.e(TAG, "Failed to get data from Liked Songs library due to authentication error with the message: ${e.message}")
+			authStateManager.addAccessTokenAuthorizationException(e)
+			createNotAuthorizedNotification()
```

Comment:
If I understand the flow properly, the webApi will only exist if the auth token was valid. Once this exception is raised, the webApi gets cleared out, and then it won't try again.
Does this mean that the AppSetting for showing the notification isn't actually needed, because this sequence will automatically show the notification just a single time when the auth becomes invalid, and then never again until reauthenticated?

Reply:
Not quite, so by default the AppSetting for showing the unauthorized notification is false. Once the user successfully authorizes for the very first time, this AppSetting is set to true to show the unauthorized notification and henceforth every time the user becomes unauthorized (whether it be the actual SpotifyWebApi client or just from an API call) the notification will be shown since they have previously logged in